rasa/core/agent.py

```diff
@@ -405,6 +405,10 @@ async def parse_message(self, message_data: Text) -> Dict[Text, Any]:
 
         """
         message = UserMessage(message_data)
+
+        if self.processor is None:
+            return {}
```

Comment:
I'm not familiar with type guards, is this described here? https://www.python.org/dev/peps/pep-0647/

Reply:
yes this is the idea

Reply:
(`isinstance()` is also a type guard mechanism)

Reply:
I've just tried rewriting the line in `is_ready()` method to `isinstance(self.processor, MessageProcessor), however mypy still complains if this is not repeated within each individual function. Also the `TypeGuard` mechanism from the PEP 647 is only available in python 3.10.

Reply:
> Also the TypeGuard mechanism from the PEP 647 is only available in python 3.10.

that's ok because we do have the `typing_extensions` module to backport it

> I've just tried rewriting the line in is_ready() method to isinstance(self.processor, MessageProcessor), however mypy still complains if this is not repeated within each individual function

how does it look like?

Reply:
This would be the code, but it still fails mypy check 😕:
```
@staticmethod
    def processor_is_not_none(processor: Optional[object]) -> TypeGuard[MessageProcessor]:
        return isinstance(processor, MessageProcessor)

    def is_ready(self) -> bool:
        """Check if all necessary components are instantiated to use agent."""
        return self.tracker_store is not None and self.processor_is_not_none(self.processor)
```

Reply:
🤔 hum ok this is underwhelming. I'd vote for disabling the mypy check (instead of adding the check for self.processor existence, because it defeats the purpose of `self.is_ready()`)

Reply:
or there is a fancy way of doing it: https://stackoverflow.com/questions/69477337/function-for-restricting-the-type-of-attributes-in-an-object-python-mypy

Reply:
(the structural sub typing one)

Reply:
This seems to imply that `MessageProcessor` should inherit from `Protocol` and that causes other mypy problems.
Also I get this PyCharm error `Class 'type' does not define '__getitem__'` when I use type the way it's used in the `check` method signature. In any case, the result is the same, the error doesn't get fixed.
I'll disable the mypy check.

Reply:
🤔 no I think it implies that we could define a `AgentReady` protocol, and the `is_ready()` method would be a type guard for the agent itself. But TBH, at this stage, it sounds a bit too much..!

Reply:
Aaa I see 👍🏼